### PR TITLE
[AArch64] Split FeatureMTE to FEAT_MTE and FEAT_MTE2.

### PR DESCRIPTION
**Pull request by labrinea**
These extensions depend on each other as follows: mte -> mte2

Currently in LLVM FeatureMTE models both FEAT_MTE and FEAT_MTE2 lumped together. However the architecture does not mandate that both need to be implemented at the same time. Therefore it makes sense to split them. This will allow Function Multiversioning to enable backend features for 'memtag' (means FEAT_MTE in FMV) which was previously impossible. I have added an override for the user visible name of FeatureMTE2 ('memtag') to preserve the old semantics for backwards compatibility with command line, target attribute and assembler directives.

### CODE REVIEW

**Comment by labrinea** [comment]
Adding the GCC folks @andrewcarlotti and @Wilco1 for visibility.

**Comment by Wilco1** [comment]
> Therefore it makes sense to split them.

I think that requires a lot more evidence - it may well be possible, but that doesn't mean it makes sense...

From the patch it is clear all CPUs that implement MTE only support MTE2 - and it seems unlikely this will change. Note the extra MTE2 instructions are only available in EL1 IIRC. So there is no use for differentiating between them for the purpose of FMV (which targets EL0). Also adding new +mte/+mte2 options on the commandline will create incompatibilities with GCC and binutils unless we did the same.

**Comment by labrinea** [comment]
> From the patch it is clear all CPUs that implement MTE only support MTE2 - and it seems unlikely this will change.

Not quite true. I didn't change those CPU entries to keep the behavior NFC, which may have been a mistake actually. The reference manual suggests that Neoverse N2, or Cortex-X2 for example only have FEAT_MTE (or at least there is no mention of FEAT_MTE2 in the documents).

> Note the extra MTE2 instructions are only available in EL1 IIRC. So there is no use for differentiating between them for the purpose of FMV (which targets EL0). 
 
This is actually a convincing argument.
 
> Also adding new +mte/+mte2 options on the commandline will create incompatibilities with GCC and binutils unless we did the same.

I am not exposing any of +mte/+mte2 on the commandline, so there shouldn't be any concerns for compatibility with GCC.

**Comment by Wilco1** [comment]
So looking at the TRMs, the cores support MTE2 or MTE3 (and MTE3 is again no user visible change), but the hardware can be configured to only support MTE1. [This is also explained here](https://developer.arm.com/documentation/ka005919/1-0).

> I am not exposing any of +mte/+mte2 on the commandline, so there shouldn't be any concerns for compatibility with GCC.

I don't quite get the code, but it seems that +mte or +mte2 may be emitted in disassembly and/or supported by the assembler?

**Comment by labrinea** [comment]
> I don't quite get the code, but it seems that +mte or +mte2 may be emitted in disassembly and/or supported by the assembler?

They are supported in assembler for the sole purpose of providing more informative error messages.

**Comment by andrewcarlotti** [comment]
I think the existing situation before FMV was already a bit confusing.

We have the command line feature `+memtag`. This enables all the instructions specified in FEAT_MTE and FEAT_MTE2. Note that the FEAT_MTE2 instructions aren't available at EL0.

With `-mcpu=neoverse-v2` in either GCC or LLVM, we enable `+memtag` by default. (I think this applies to all CPUs with memtag support). The actual CPU always supports the FEAT_MTE instructions, but it can only support FEAT_MTE2 if other parts of the system include support for it. So this would suggest that the `+memtag` flag corresponds to FEAT_MTE, with the non-EL0 FEAT_MTE2 instructions requiring the user to check that they're actually supported by the target system.

In contrast, HWCAP_MTE is enabled in the Linux kernel only when support exists for FEAT_MTE2. This means that the HWCAP doesn't directly match the command line flag.

I think this difference would be ok in practice, since I can't imagine any scenario in which it would be beneficial to choose a `+memtag` version of a function when FEAT_MTE2 is not supported. However, if we include `+memtag` in FMV with this discrepancy, then we do have to be careful to avoid indirection elimination optimisations that assume that a function with `+memtag` enabled (via the command line or attributes) will always call a `"memtag"` function version at runtime.

File: clang/lib/Basic/Targets/AArch64.cpp [hunk 1]
```diff
@@ -1059,7 +1059,7 @@ bool AArch64TargetInfo::handleTargetFeatures(std::vector<std::string> &Features,
       HasFullFP16 = true;
       HasFP16FML = true;
     }
-    if (Feature == "+mte")
+    if (Feature == "+mte2")
```
Comment (tmatheson-arm) [inline clang/lib/Basic/Targets/AArch64.cpp:1062]: Won't this break existing IR with `target-features="mte"`?

Comment (tmatheson-arm) [inline clang/lib/Basic/Targets/AArch64.cpp:1062]: This breaks backwards compatibility of IR. _If_ you want to go ahead with this patch (which is fundamentally just a change to the internal representation of features) then I would suggest breaking backwards compatibility by having `target-feature="+mte"` mean FEAT_MTE and `target-feature="+mte2"` mean FEAT_MTE2. Then if anyone happens to be using MTE2 instructions in inline assembly with `target-feature="+mte"` they will get an error, but for the vast majority of cases, the change from FEAT_MTE2 to FEAT_MTE for `target-feature="+mte"` will have no effect.

**Comment by Wilco1** [comment]
> I think this difference would be ok in practice, since I can't imagine any scenario in which it would be beneficial to choose a +memtag version of a function when FEAT_MTE2 is not supported. However, if we include +memtag in FMV with this discrepancy, then we do have to be careful to avoid indirection elimination optimisations that assume that a function with +memtag enabled (via the command line or attributes) will always call a "memtag" function version at runtime.

Isn't the latter true for many features? A feature may be enabled on the command-line, but the kernel may turn it off.

Also HWCAP2_MTE is correct in the sense that it is only enabled when MTE can be used on the system. There is no point in selecting an MTE function if the kernel or system does not support MTE anyway...

**Pull request closed without merging**